### ISSUE TITLE
Vi mode defaults to block cursor always

### ISSUE THREAD

**Issue by bool3max** [fish-shell/fish-shell#8635]
`fish --version`: `3.3.1`
`kitty --version`: `0.24.1`

I am running `fish` with Vi-mode keybindings

```
set -g fish_key_bindings fish_vi_key_bindings
```
When first entering `fish`, I am placed in insert mode as expected, and the cursor is a regular thin caret: 

![image](https://user-images.githubusercontent.com/12991998/149414467-850db1cf-848e-4eeb-a2fb-8b62ec85ffa6.png)

Going into normal mode results in the cursor switch to a block, as expected: 

![image](https://user-images.githubusercontent.com/12991998/149414571-d6ecc4c2-f5c3-4d1d-9b0b-15fee6151601.png)

However, from then on, switching back to insert mode **does not** result in the cursor style going back to a thin caret: 

![image](https://user-images.githubusercontent.com/12991998/149414671-97bfa463-4f35-4373-888b-1821e3f22b6f.png)

---

As the kitty terminal supports [shell-specific integration](https://sw.kovidgoyal.net/kitty/shell-integration/#shell-integration) (including that for fish), I am unsure if this is a fish-specific problem or a kitty issue. If it is unrelated to fish, please feel free to close the issue.

Note that disabling shell-integration in kitty (via `shell_integration none` in the config) results in only one cursor style (that set by the terminal), as expected.

**Comment by faho** [comment]
What are $fish_cursor_default and $fish_cursor_insert set to?

**Comment by page-down** [comment]
In my environment, `fish_cursor_default`, `fish_cursor_insert` are not set.

The related shell integration code is:

```fish
function _ksi_bar_cursor --on-event fish_prompt
    printf "\e[5 q"
end
function _ksi_block_cursor --on-event fish_preexec
    printf "\e[2 q"
end
```

**Comment by faho** [comment]
Okay, you've not told fish any cursor shapes.

There are two things at play here:

1. Kitty apparently redoes our cursor shape logic
2. When not given a cursor shape, fish will fall back to $fish_cursor_unknown, which defaults to block

So kitty will set the cursor shape before and after a command, but fish will also set the cursor both then and on mode switch to block every time.

We should probably stop that second bit and kitty should probably stop that first bit.

You should simply set $fish_cursor_insert to "line".

**Comment by page-down** [comment]
Thanks for the explanation, I'll see if I can make improvements regarding the current situation.

What is the best way to change the normal (non-vi mode) cursor shape?

**Comment by faho** [comment]
>What is the best way to change the normal (non-vi mode) cursor shape?

Set it in your terminal. There's no real need for fish to do it because once you don't have modal bindings that change the cursor depending on the mode you're in you just have a constant cursor shape.

**Comment by page-down** [comment]
Is it possible to use `fish_cursor_insert` for "fish_default_key_bindings"?
If user want to use `line`  only in fish emacs-like (insert) mode, but do not want to change the terminal fallback default shape, since it will affect other programs.

Honestly, I'd like to get rid of that event-based code. `printf "\e[5 q"`

**Comment by faho** [comment]
>but do not want to change the terminal fallback default shape, since it will affect other programs.

#4656 asks for a "$fish_cursor_external", which I feel is nicer.

**Comment by iHnR** [comment]
I had the same issue (also with kitty terminal). It seems a workaround is to place the cursor shape commands outside the `fish_user_key_bindings` function. The config below works for me.  
The only issue that remains with this is that a newline is randomly inserted on some vi commands some of the time, but I believe this is due to my starship prompt having two lines.

```fish
# Only run this in interactive shells
if status is-interactive

  # I'm trying to grow a neckbeard
  # fish_vi_key_bindings
  # Set the cursor shapes for the different vi modes.
  set fish_cursor_default     block      blink
  set fish_cursor_insert      line       blink
  set fish_cursor_replace_one underscore blink
  set fish_cursor_visual      block

  function fish_user_key_bindings
    # Execute this once per mode that emacs bindings should be used in
    fish_default_key_bindings -M insert
    fish_vi_key_bindings --no-erase insert
  end
end
```

**Comment by faho** [comment]
Okay it seems to me like this should be fixed by #4656 (and issues with kitty's shape changing integration are of course issues in kitty).
